### PR TITLE
Client support for offline file storage

### PR DESCRIPTION
**Pull request by wtrocki**
## Motivation

Allow to store files in temporary storage if internet is not available. 

## Usage 
```
cd raincatcher-demo-mobile
npm install --save fh-wfm-file@0.4.0-pre.2
```
ping @col1985 

See https://issues.jboss.org/browse/RAINCATCH-636 for verification steps.

### CODE REVIEW

**Comment by col1985** [comment]
@wtrocki this is working well in current solution. One issue is that we need to return the metaData object with reference to the file in gridFS in the done callback. 

I have also updated `createFile` to accept fileName. Can you please review and publish as a pre-release?

**Comment by wtrocki** [comment]
@col1985 Reviewed your PR and integrated changes into this PR.

File: lib/client/file-client/FileClient.js [hunk 1]
```diff
@@ -81,6 +84,24 @@ FileClient.prototype.createFile = function(file) {
 
 
 /**
+ * Add file to upload queue. File would be uploaded depending on internet connectivity.
+ *
+ * @param file {userId, fileURI, options, dataUrl}
+ * @returns {*}
+ */
+FileClient.prototype.sheduleFileToBeUploaded = function(file) {
```
Comment (JameelB) [inline lib/client/file-client/FileClient.js]: small typo here: `scheduleFileToBeUploaded`

File: lib/client/mediator-subscribers/create.js [hunk 2]
```diff
@@ -22,17 +22,16 @@ module.exports = function createFileSubscriber(fileEntityTopics, fileClient) {
...
     //If no file is passed, can't create one
     if (!_.isPlainObject(fileToCreate)) {
       return self.mediator.publish(fileCreateErrorTopic, new Error("Invalid Data To Create A File."));
     }
-
-    fileClient.createFile(fileToCreate)
-    .then(function(response) {
-      self.mediator.publish(fileEntityTopics.getTopic(CONSTANTS.TOPICS.CREATE, CONSTANTS.DONE_PREFIX, parameters.topicUid), response);
+  
+    fileClient.sheduleFileToBeUploaded(fileToCreate).then(function(result) {
```
Comment (JameelB) [inline lib/client/mediator-subscribers/create.js]: fix typo: `scheduleFileToBeUploaded`

File: lib/client/offline-manager/queue.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+queue.prototype.getItemList = function() {
+  return this.queueData;
+};
+
+/**
+ * @param {object} item meta data model
+ */
+queue.prototype.addItem = function(item) {
+  this.queueData.push(item);
+  // TODO save queue as array of items
```
Comment (JameelB) [inline lib/client/offline-manager/queue.js]: remove `todo` comment

**Review by JameelB — APPROVED** [review]
Looks good to me, small comments below

**Review by witmicko — APPROVED** [review]
LGTM bar few typos as pointed out by @JameelB,
any chance you could add debug logger or replace console logs entirely?

**Comment by wtrocki** [comment]
Resolved all issues. Reviewed by @JameelB 
Verified by @witmicko

File: lib/client/offline-manager/index.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  return this.uploadQueue.addItem(object);
+};
+
+FileQueueManager.prototype.startProcessingUploads = function() {
+  var queueItems = this.uploadQueue.restoreData().getItemList();
+  var promise;
+  var self = this;
+  if (queueItems && queueItems.length > 0) {
+    console.log("Processing offline file queue. Number of items to save: ", queueItems.length);
+    promise = this.saveFile(queueItems[0]).then(function() {
```
Comment (vchepeli) [inline lib/client/offline-manager/index.js]: here you should call this.saveFile(_.noop) otherwise you will upload first file`queueItems[0]` twice

**Pull request merged**